### DISCUSSION TITLE
a gui that sends and receives ggwave data in python.

### DISCUSSION THREAD

**Discussion by denizsincar29** [ggerganov/ggwave#87] [Show and tell]
hello guys. Long ago I made a repository https://github.com/denizsincar29/data_over_sound
it is a simple, screenreader accessible gui ggwave sender and receiver. I made it because the ios waver is not accessible with voiceover screenreader for blind people.
take and play, hope you enjoyed.
to do:
choosing input and output devices via gui, don't know how to beautifully do it.
improve the file sending protocol, currently it is the most basic one.
please guys if you find it interesting, feel free to pullrequest. especially with beautifying the gui. i'm blind, I don't see the gui. I just navigate with a screenreader.
Thank you.

**Comment by denizsincar29** [comment]
@ggerganov Maybe you could add this to third party examples?

**Reply by ggerganov** [reply]
Hi @denizsincar29,
I moved the discussion into the "Show and tell" category and pinned it at the top.
Thanks for making this!

**Reply by denizsincar29** [reply]
hy Georgy. How to fork the ggwave repo? I didn't se the fork button surprisingly, but it is there in all other repos. I want to do a little change in the ggwave python package.

**Reply by ggerganov** [reply]
It's in the upper right corner of the page - to the right of "Watch" and to the left of "Star"

**Reply by denizsincar29** [reply]
ah, I found it with the find function of my screenreader for blind.

**Comment by denizsincar29** [comment]
@ggerganov can you test my gui app and see why my app doesn't recognize monotonic payload length 16 even if I sent a command: /p 11 16, which meens protocol 11 (mt fastest), length 16.?
thank you.

**Comment by denizsincar29** [comment]
Also @ggerganov please fix ggwave to install correctly to python 3.11 issue #89 or everytime I need to do it with python 3.10.

**Comment by denizsincar29** [comment]
Hey guys, here's an update. I've redone my ggwave data over sound repo, and it's now using CLI again, because I've been exhausted with this gui library. Also I completely remove file sharing and want to reimplement it. Here's my idea:
For short terms, i use tx for transmitting computer and rx for receiving device.
1. Tx sends a handshake to rx with number of chunks and a hashsum.
2. rx replies with a message that says: OK, start sending!
3. TX starts to send the chunks with protocol №2, which is fastest dual tone. One by one, each chunk is separated with 200 ms pause. It starts with an 4 byte index or whatever metadata needed, and 128 bytes of file chunk.
4. The sender sends some marker that it's finished transmitting.
5. The receiver sends either "OK" or "didnt_hear:" and number of chunks comma separated. E.g.: "didnt_hear:3,5" means that receiver didn't hear 3rd and 5th chunk.
6. If didn't hear, the sender sends that chunks again untill receiver heard. Receiver sends OK or ERR if hash sum is bad.
File sending is not ment to send megabites of data. 1 kb will go about 80 seconds, so send text files or other small pieces of data. This is done for experimental purposes.
Anyone knows how to send bytes, not string, in ggwave python lib?
